cmd/gapic-showcase/qualifier/scenario.go

```diff
@@ -0,0 +1,359 @@
+		}
+		fileTypes, err = yamlDocTypes(content)
+		if err != nil {
+			return fileTypes, err
+		}
+	}
+	fileTypes = append(fileTypes, "scenario")
+	return fileTypes, err
+}
+
+// yamlDocTypes returns a list of the document types encountered in the (possibly multi-document)
+// YAML payload `fileContent`.
+func yamlDocTypes(fileContent []byte) (docTypes []string, err error) {
+	type docSchema struct {
+		Type string
+	}
+
+	decoder := yaml.NewDecoder(bytes.NewReader(fileContent))
+	for {
+		doc := &docSchema{Type: "(UNKNOWN)"}
+		err = decoder.Decode(doc)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			err = fmt.Errorf("error decoding file: %w", err)
+			trace.Trace(err)
+			return
+		}
+		docTypes = append(docTypes, doc.Type)
+	}
+	return docTypes, nil
+}
+
+// createSandbox creates a sandbox directory within which all the generation and checks for this
+// scenario will be run.
+func (scenario *Scenario) createSandbox() (err error) {
+	if scenario.sandbox, err = ioutil.TempDir("", fmt.Sprintf("showcase-qualify.%s.%s.", scenario.timestamp, scenario.name)); err != nil {
+		err := fmt.Errorf("could not create sandbox: %w", err)
+		trace.Trace(err)
+		return err
+	}
+
+	err = scenario.copyFiles(scenario.filesByType["scenario"])
+	if err != nil {
+		err = fmt.Errorf("could not copy scenario files: %w", err)
+		trace.Trace(err)
+		return err
+	}
+
+	if err = scenario.getIncludes(scenario.filesByType["include"]); err != nil {
+		err = fmt.Errorf("could not copy included schema files: %w", err)
+		trace.Trace(err)
+		return err
+	}
+
+	trace.Trace("created sandbox")
+	return nil
+}
+
+// getIncludes processes include files (whose names start with `includeFilePrefix`) by replacing
+// them with the file or directory referenced in the file's contents. The only currently supported
+// type of reference is to either an entry in `scenario.schemaBox`, or a directory that resolves to
+// one or more entries in `scenario.schemaBox`. If an include file does not resolve to any entries
+// in `scenario.schemaBox`, this function returns an error.
+//
+// If there's need in the future, we could potentially grow to support fetching files or directories
+// from public repositories on the web.
+func (scenario *Scenario) getIncludes(includes []string) error {
+	for _, includeFile := range includes {
+
+		// The following guards against includeFilePrefix being present elsewhere in
+		// includeFile. Otherwise, we could simply use
+		//
+		//   `dstPath := strings.Replace(includeFile, includeFilePrefix, "")`
+		prefixStartIdx := strings.LastIndex(includeFile, includeFilePrefix)
+		prefixEndIdx := prefixStartIdx + len(includeFilePrefix)
+		if prefixStartIdx < 0 || prefixEndIdx >= len(includeFile) {
+			msg := fmt.Sprintf("logic error: start %d, end %d outside of range [0, %d] for %q", prefixStartIdx, prefixEndIdx, len(includeFile), includeFile)
+			trace.Trace(msg)
+			panic(msg)
+		}
+		dstPath := includeFile[0:prefixStartIdx] + includeFile[prefixEndIdx:]
```

Comment:
> Isn't your "w/o ReplaceAll" exactly what I had, except you inlined prefixEndIdx?

Yes, that one line is, but the whole suggestion is not the same. The suggestion as a whole makes changes that preclude the need for that variable to be explicitly declared, hence why I inlined it. The entire suggestion offers a more concise approach than what was there when I reviewed it, as opposed to what is there [now](https://github.com/vchudnov-g/gapic-showcase/blob/fe47602bdb5c1a04f8cd6a75ece65032927229f1/cmd/gapic-showcase/qualifier/scenario.go#L260-L267) which is definitely an improvement on the original. 

I take it that using strings.Contains rather than checking the index and inlining the prefixEndIdx variable are not preferred?